pandas/core/nanops.py

```diff
@@ -829,9 +830,42 @@ def _checked_add_with_arr(arr, b):
 
     Raises
     ------
-    OverflowError if any x + y exceeds the maximum int64 value.
+    OverflowError if any x + y exceeds the maximum or minimum int64 value.
     """
-    if (np.iinfo(np.int64).max - b < arr).any():
-        raise OverflowError("Python int too large to "
-                            "convert to C long")
+    # For performance reasons, we broadcast 'b' to the new array 'b2'
+    # so that it has the same size as 'arr'.
+    if _np_version_under1p10:
+        if lib.isscalar(b):
+            b2 = np.empty(arr.shape)
+            b2.fill(b)
+        else:
+            b2 = b.copy()
+    else:
+        b2 = np.broadcast_to(b, arr.shape)
+
+    # For each element in 'arr' and its corresponding element in 'b2',
+    # we check the sign of the element in 'b2'. If it is positive, we
+    # then check whether its sum with the element in 'arr' exceeds
+    # np.iinfo(np.int64).max. If so, we have an overflow error. If it
+    # it is negative, we then check whether its sum with the element in
+    # 'arr' exceeds np.iinfo(np.int64).min. If so, we have an overflow
+    # error as well.
+    #
+    # gh-14324: The implementation below takes advantage of NumPy
+    # broadcasting to check all of these elements at once. Although
+    # it is somewhat convoluted, it has much better performance than
+    # say iterating through all elements in both 'arr' and 'b2'.
```

Comment:
Maybe instead of this comment, I would clarify that the checks have to be performed separately for negative/positive overflow for performance reasons

(the fact that iterating over an array in pure python is not performant should be rather obvious to developers I think)